### PR TITLE
Replace Cgroup Parent and Name fields by CgroupsPath

### LINKED ISSUE #396: runC ignores passed in cgroupsPath
Current version of runC(0.0.4) ignore passed in cgroupsPath.
According to the spec, cgroupsPath (https://github.com/opencontainers/specs/blob/master/runtime_config_linux.go#L31) can be passed in, and implementations will join that cgroup(if it exists).

### PR DESCRIPTION
**Pull request by mlaventure**
Fixes #396

Closes #397 
Closes #399 

Signed-off-by: Kenfe-Mickael Laventure <mickael.laventure@gmail.com>

---

Since #397 and #399 haven't been touched in a month, I've made this PR.

This hopefully includes all decisions following the discussions in the aforementioned PRs.

However, since #466 was merged, we may need a better name than CgroupsPath now that we have a "Paths" member, but I couldn't come up with a better name. Open to suggestions.

### CODE REVIEW

File: libcontainer/cgroups/fs/apply_raw.go [hunk 1]
```diff
@@ -273,45 +272,17 @@ func getCgroupPath(c *configs.Cgroup) (string, error) {
 	return d.path("devices")
 }
 
-// pathClean makes a path safe for use with filepath.Join. This is done by not
-// only cleaning the path, but also (if the path is relative) adding a leading
-// '/' and cleaning it (then removing the leading '/'). This ensures that a
-// path resulting from prepending another path will always resolve to lexically
-// be a subdirectory of the prefixed path. This is all done lexically, so paths
-// that include symlinks won't be safe as a result of using pathClean.
-func pathClean(path string) string {
```
Comment (cyphar) [inline libcontainer/cgroups/fs/apply_raw.go]: **NACK** on removing this. It fixes a security vulnerability. Please use this function to clean paths, not `filepath.Clean`.

File: spec.go [hunk 2]
```diff
@@ -448,15 +448,48 @@ func createLibcontainerMount(cwd, dest string, m specs.Mount) *configs.Mount {
 	}
 }
 
+// pathClean makes a path safe for use with filepath.Join. This is done by not
+// only cleaning the path, but also (if the path is relative) adding a leading
+// '/' and cleaning it (then removing the leading '/'). This ensures that a
+// path resulting from prepending another path will always resolve to lexically
+// be a subdirectory of the prefixed path. This is all done lexically, so paths
+// that include symlinks won't be safe as a result of using pathClean.
+func pathClean(path string) string {
```
Comment (cyphar) [inline spec.go]: Ah, okay. I see what you've done. I'm not too sure why you've moved this. Docker binds directly into the `libcontainer` stuff, so this change will break security **in Docker**.

File: libcontainer/cgroups/fs/apply_raw.go (hunk 1 above)
Comment (mlaventure) [inline libcontainer/cgroups/fs/apply_raw.go]: I should have noted that, yes.

What about making it a public method within the libcontainer/utils package?

Comment (cyphar) [inline libcontainer/cgroups/fs/apply_raw.go]: SGTM, and I actually think I mentioned that in the commit message as a "TODO".

**Comment by hqhq** [comment]
I think we are waiting for https://github.com/opencontainers/specs/pull/270
It's gonna be a bit tough for systemd cgroup, do you have any idea how we should handle cgroupName for systemd cgroup? Just split the path seems not enough.

**Comment by mlaventure** [comment]
Ah, I missed opencontainers/specs#270 somehow, sorry.

For systemd, I tried to get the same result as before. I'll double check on a vm tomorrow in case I missed something (I basically runc a busybox with `cat /proc/self/cgroup` and got the expected result with my solution).

 Do we have a systemd expert in the contributors?

File: libcontainer/cgroups/fs/apply_raw.go [hunk 3]
```diff
@@ -273,45 +273,17 @@ func getCgroupPath(c *configs.Cgroup) (string, error) {
...
-	c.Parent = pathClean(c.Parent)
-
 	return &cgroupData{
-		root:   root,
-		parent: c.Parent,
-		name:   c.Name,
-		config: c,
-		pid:    pid,
+		root:        root,
+		cgroupspath: libcontainerUtils.CleanPath(c.CgroupsPath),
```
Comment (cyphar) [inline libcontainer/cgroups/fs/apply_raw.go]: Also, why isn't this just called `path`? After all, the struct is called `cgroupData`.

**Comment by cyphar** [comment]
I'm not a systemd expert (@LK4D4 is the closest we have AFAIK), but I've dealt with my fair share of systemd bugs recently.

TransientUnits require a slice and a unit name. They are then concatenated. AFAICS `filepath.Split` is fine as long as the path has been `Clean`d properly.

**Comment by hqhq** [comment]
It's more like a strategy problem, something like, if we demand cgroupsPath to be absolute (as specs required), if user use `cgroupsPath=/a.slice/b.slice/mycgroup.scope`, `filepath.Split` won't work because `/a.slice/b.slice` is not a valid slice name, and what if some subsystems don't have the path `/a.slice/b.slice` should we fail or create this path? And how should systemd `cgroupsPath` work with `--cgroup-parent` on Docker side?
We need to think through these questions before process.

**Comment by cyphar** [comment]
@hqhq Well, the slice should be `a.slice/b.slice` in your example (`systemd` scopes slices to `/`, unless you do something like add a `-` in your slice name). The "must be absolute" rules are because of systemd interactions with `cgroupfs` (something I know about quite viscerally).

**Comment by mlaventure** [comment]
On docker side, we can just concatenate `Parent` and `Name` and use the result to fill up `cgroupsPat h`.

For systemd, I just realize that runc always uses fs regardless of whether systemd is installed on the localhost or not. We only do the detection when doing the test. I'll have to redo my testing.

**Comment by mlaventure** [comment]
So, even master does not currently play nice with systemd (at least in my ubuntu 15.10 vm). It fails with the following
> FATA[0000] Container start failed: [9] System error: Invalid slice name /user.slice/user-1000.slice/session-c1.scope 


`createCgroupConfig` will use the path as read from within `/proc/self/group` which result in an invalid path name format as far as systemd is concerned (e.g. `/user.slice/user-1000.slice/session-c1.scope` which should be AFAIC something like `user-1000.slice`). The scope itself is another issue, I guess it's what we would consider to be `Name` currently. But systemd as the concept of `ScopePrefix` (`session` in the previous example) that we never fill. I guess we should probably take the default from `/proc/self/cgroup` (that would be `session` if we stick to my example).

Since this is 100% systemd specific, `GetThisCgroupDir` should probably moved within the Cgroup Managers.

**Comment by mrunalp** [comment]
For systemd, we create the scope using the dbus APIs and then manipulation is done by fs. Creating the scope using dbus is important because systemd does more than just create the paths.

**Comment by mlaventure** [comment]
@mrunalp , yes we do. But by default with use the path as retrieved from `/proc/self/cgroup` when calling the dbu API.

The issue is that this path is in an invalid format. It would need to either be converted to an appropriate format or we should instead use `system.slice` if nothing was provided by the user.

**Comment by cyphar** [comment]
@mlanventure Can you please share the command you used on `master`? All of these work for me:

```
$ ./bundles/latest/binary/docker run alpine cat /proc/self/cgroup
11:cpuset:/system.slice/docker-f34a95b1920a7160c1656f4651ebde7a2e3b14147f67c3249f55e37117bf6204.scope
10:memory:/system.slice/docker-f34a95b1920a7160c1656f4651ebde7a2e3b14147f67c3249f55e37117bf6204.scope
9:perf_event:/system.slice/docker-f34a95b1920a7160c1656f4651ebde7a2e3b14147f67c3249f55e37117bf6204.scope
8:freezer:/system.slice/docker-f34a95b1920a7160c1656f4651ebde7a2e3b14147f67c3249f55e37117bf6204.scope
7:blkio:/system.slice/docker-f34a95b1920a7160c1656f4651ebde7a2e3b14147f67c3249f55e37117bf6204.scope
6:devices:/system.slice/docker-f34a95b1920a7160c1656f4651ebde7a2e3b14147f67c3249f55e37117bf6204.scope
5:net_cls,net_prio:/system.slice/docker-f34a95b1920a7160c1656f4651ebde7a2e3b14147f67c3249f55e37117bf6204.scope
4:hugetlb:/system.slice/docker-f34a95b1920a7160c1656f4651ebde7a2e3b14147f67c3249f55e37117bf6204.scope
3:pids:/
2:cpu,cpuacct:/system.slice/docker-f34a95b1920a7160c1656f4651ebde7a2e3b14147f67c3249f55e37117bf6204.scope
1:name=systemd:/system.slice/docker-f34a95b1920a7160c1656f4651ebde7a2e3b14147f67c3249f55e37117bf6204.scope
$ ./bundles/latest/binary/docker run --cgroup-parent=test.slice alpine cat /proc/self/cgroup
11:cpuset:/test.slice/docker-c3d52f73a3b86908c3dc57674ca3143a2274ba3d4e1e2647afaa115e2aacaf60.scope
10:memory:/test.slice/docker-c3d52f73a3b86908c3dc57674ca3143a2274ba3d4e1e2647afaa115e2aacaf60.scope
9:perf_event:/test.slice/docker-c3d52f73a3b86908c3dc57674ca3143a2274ba3d4e1e2647afaa115e2aacaf60.scope
8:freezer:/test.slice/docker-c3d52f73a3b86908c3dc57674ca3143a2274ba3d4e1e2647afaa115e2aacaf60.scope
7:blkio:/test.slice/docker-c3d52f73a3b86908c3dc57674ca3143a2274ba3d4e1e2647afaa115e2aacaf60.scope
6:devices:/test.slice
5:net_cls,net_prio:/test.slice/docker-c3d52f73a3b86908c3dc57674ca3143a2274ba3d4e1e2647afaa115e2aacaf60.scope
4:hugetlb:/test.slice/docker-c3d52f73a3b86908c3dc57674ca3143a2274ba3d4e1e2647afaa115e2aacaf60.scope
3:pids:/
2:cpu,cpuacct:/test.slice/docker-c3d52f73a3b86908c3dc57674ca3143a2274ba3d4e1e2647afaa115e2aacaf60.scope
1:name=systemd:/test.slice/docker-c3d52f73a3b86908c3dc57674ca3143a2274ba3d4e1e2647afaa115e2aacaf60.scope
$ ./bundles/latest/binary/docker run --cgroup-parent=test-a-b-c.slice alpine cat /proc/self/cgroup
11:cpuset:/test-a-b-c.slice/docker-e5fc84ab93f9cd3c2a5308d7072c91474b3e01abceb8524746bd6223964af059.scope
10:memory:/test.slice/test-a.slice/test-a-b.slice/test-a-b-c.slice/docker-e5fc84ab93f9cd3c2a5308d7072c91474b3e01abceb8524746bd6223964af059.scope
9:perf_event:/test-a-b-c.slice/docker-e5fc84ab93f9cd3c2a5308d7072c91474b3e01abceb8524746bd6223964af059.scope
8:freezer:/test-a-b-c.slice/docker-e5fc84ab93f9cd3c2a5308d7072c91474b3e01abceb8524746bd6223964af059.scope
7:blkio:/test.slice/test-a.slice/test-a-b.slice/test-a-b-c.slice/docker-e5fc84ab93f9cd3c2a5308d7072c91474b3e01abceb8524746bd6223964af059.scope
6:devices:/test.slice
5:net_cls,net_prio:/test-a-b-c.slice/docker-e5fc84ab93f9cd3c2a5308d7072c91474b3e01abceb8524746bd6223964af059.scope
4:hugetlb:/test-a-b-c.slice/docker-e5fc84ab93f9cd3c2a5308d7072c91474b3e01abceb8524746bd6223964af059.scope
3:pids:/
2:cpu,cpuacct:/test.slice/test-a.slice/test-a-b.slice/test-a-b-c.slice/docker-e5fc84ab93f9cd3c2a5308d7072c91474b3e01abceb8524746bd6223964af059.scope
1:name=systemd:/test.slice/test-a.slice/test-a-b.slice/test-a-b-c.slice/docker-e5fc84ab93f9cd3c2a5308d7072c91474b3e01abceb8524746bd6223964af059.scope
```

Which all work (although the last one causes warnings about missing kernel features because we aren't expanding paths the way that systemd does with `-` becoming sub-slices -- which is definitely a bug). What wouldn't work is something with `/`s inside (like `--cgroup-parent=a.slice/b.slice`), and it shouldn't IMO.

Really, we should rethink how we're dealing with systemd's transient units IMHO. I can take a look at this later.

**Comment by cyphar** [comment]
Actually, the output of `--cgroup-parent=test-a-b-c.slice` is blatantly incorrect for cgroups the systemd doesn't support (which I should work on fixing).

**Comment by mlaventure** [comment]
@cyphar `docker` works because it calls `libcontainer.New()` directly. I am referring to the runc binary itself. It does not provide a way to use anything but cgroup fs. When adding the check for systemd you end up with the error I mentioned above.

**Comment by cyphar** [comment]
#511 actually solves the above issue I describe about incorrect cgroup paths.

@mlaventure Can you share the code you used to add the systemd check?

**Comment by mlaventure** [comment]
@cyphar I've used this:
```Diff
diff --git a/libcontainer/factory_linux.go b/libcontainer/factory_linux.go
index ae5db9c..0700355 100644
--- a/libcontainer/factory_linux.go
+++ b/libcontainer/factory_linux.go
@@ -114,7 +114,9 @@ func New(root string, options ...func(*LinuxFactory) error) (Factory, error) {
                CriuPath:  "criu",
        }
        InitArgs(os.Args[0], "init")(l)
-       Cgroupfs(l)
+       if l.NewCgroupsManager != nil {
+               Cgroupfs(l)
+       }
        for _, opt := range options {
                if err := opt(l); err != nil {
                        return nil, err
diff --git a/utils.go b/utils.go
index baffef1..8015d39 100644
--- a/utils.go
+++ b/utils.go
@@ -9,6 +9,7 @@ import (
 
        "github.com/codegangsta/cli"
        "github.com/opencontainers/runc/libcontainer"
+       "github.com/opencontainers/runc/libcontainer/cgroups/systemd"
        "github.com/opencontainers/runc/libcontainer/configs"
        "github.com/opencontainers/specs"
 )
@@ -104,7 +105,14 @@ func loadFactory(context *cli.Context) (libcontainer.Factory, error) {
        if err != nil {
                return nil, err
        }
-       return libcontainer.New(abs, libcontainer.Cgroupfs, func(l *libcontainer.LinuxFactory) error {
+       var newCgroupFunc func(l *libcontainer.LinuxFactory) error
+       if systemd.UseSystemd() {
+               newCgroupFunc = libcontainer.SystemdCgroups
+       } else {
+               newCgroupFunc = libcontainer.Cgroupfs
+       }
+
+       return libcontainer.New(abs, newCgroupFunc, func(l *libcontainer.LinuxFactory) error {
                l.CriuPath = context.GlobalString("criu")
                return nil
        })
```

**Comment by hqhq** [comment]
@mlaventure Yeah I've done this before https://github.com/opencontainers/runc/issues/325
Haven't dig into this but seems we don't have plan to support systemd cgroup in runc yet.

File: libcontainer/configs/cgroup_unix.go [hunk 4]
```diff
@@ -11,15 +11,15 @@ const (
 )
 
 type Cgroup struct {
-	Name string `json:"name"`
-
-	// name of parent cgroup or slice
-	Parent string `json:"parent"`
+	// CgroupsPath specifies the path to cgroups that are created and/or joined by the container.
+	// The path is assumed to be relative to the host system cgroup mountpoint.
+	CgroupsPath string `json:"cgroupsPath"`
```
Comment (crosbymichael) [inline libcontainer/configs/cgroup_unix.go]: Would just call this `Path`

File: libcontainer/cgroups/fs/apply_raw.go [hunk 5]
```diff
@@ -94,11 +95,10 @@ func getCgroupRoot() (string, error) {
 }
 
 type cgroupData struct {
-	root   string
-	parent string
-	name   string
-	config *configs.Cgroup
-	pid    int
+	root        string
+	cgroupspath string
```
Comment (crosbymichael) [inline libcontainer/cgroups/fs/apply_raw.go]: `path`

**Comment by mlaventure** [comment]
* Rebased
* Reverted the changes to systemd, since it is not exposed via runc anyway.
* Remove "cgroups" from the field names

File: spec.go [hunk 6]
```diff
@@ -326,13 +327,22 @@ func createLibcontainerMount(cwd string, m specs.Mount) *configs.Mount {
...
+		myCgroupPath, err = cgroups.GetThisCgroupDir("devices")
+		if err != nil {
+			return nil, err
+		}
 	}
+
 	c := &configs.Cgroup{
-		Name:      name,
-		Parent:    myCgroupPath,
+		Path:      filepath.Join(myCgroupPath, name),
```
Comment (crosbymichael) [inline spec.go]: I don't think you need to do a join here with name right?

File: libcontainer/configs/cgroup_unix.go [hunk 7]
```diff
@@ -13,13 +13,18 @@ const (
 type Cgroup struct {
 	Name string `json:"name"`
```
Comment (crosbymichael) [inline libcontainer/configs/cgroup_unix.go]: Can you add a comment saying these fields are deprecated and omitempty to `Name` and `Parent` ?

File: libcontainer/cgroups/fs/apply_raw.go [hunk 8]
```diff
@@ -267,45 +268,23 @@ func getCgroupPath(c *configs.Cgroup) (string, error) {
...
 		return nil, err
 	}
 
-	// Clean the parent slice path.
-	c.Parent = pathClean(c.Parent)
+	if (c.Name != "" || c.Parent != "") && c.Path != "" {
+		return nil, fmt.Errorf("cgroup: either Path or Name and Parent should be used")
+	}
 
 	return &cgroupData{
```
Comment (crosbymichael) [inline libcontainer/cgroups/fs/apply_raw.go]: If cgroup data is not used between systemd and fs we could remove the name and parent fields here and only hold the joined "path" if that makes sense

File: spec.go (hunk 6 above)
Comment (mlaventure) [inline spec.go]: it depends, if I have a bundle named `busybox` and set my cgroup path to be `foo/bar`

Do we want to use `<cgroup-mount-point>/foo/bar/busybox` or `<cgroup-mount-point>/foo/bar` when setting the rules?

Atm the code produce the former; if we want the later, I'm not sure what is the use for name.

Comment (crosbymichael) [inline spec.go]: ok

**Comment by mlaventure** [comment]
* Added deprecation comment to Name and Path along with json omitempty tag
* Removed name and parent fields from cgroupfs cgroupData type

File: libcontainer/utils/utils.go [hunk 9]
```diff
@@ -54,3 +55,27 @@ func WriteJSON(w io.Writer, v interface{}) error {
...
+	// "/../../../../../" which can cause lots of issues).
+	path = filepath.Clean(path)
+
+	// If the path isn't absolute, we need to do more processing to fix paths
+	// such as "../../../../<etc>/some/path". We also shouldn't convert absolute
+	// paths to relative ones.
+	if !filepath.IsAbs(path) {
+		path = filepath.Clean(string(os.PathSeparator) + path)
+		// This can't fail, as (by definition) all paths are relative to root.
+		path, _ = filepath.Rel(string(os.PathSeparator), path)
```
Comment (crosbymichael) [inline libcontainer/utils/utils.go]: This is causing inconsistencies whenever i specify `/something` vs `something` as the cgroup path

**Comment by crosbymichael** [comment]
LGTM

**Comment by LK4D4** [comment]
LGTM

**Comment by cyphar** [comment]
It turns out that `256f3a8ebc114b01c1630571c33d813a0612ea5c` (Add support for CgroupsPath field) breaks cgroups if you vendor it into Docker. I'm going to see what I can do to fix this ...

**Pull request merged**